File: framethrower/branches/v1.6.0/js/testSyntaxParser/testsyntax.js
function testFile(filename) {
	//helper functions
	
	var scPatternMatch = {};
	
	function children (xml) {
		return xpath("*", xml);
	}
	

	function messagesEqual (message1, message2) {
		var returnVal = true;
		var propCount = 0;
		
		forEach(message1, function(property, pname) {
			if(message2[pname] == undefined) {
				returnVal = false;
				return;
			}
			if(message2[pname].scPatternMatch == scPatternMatch) {
				message2[pname].callBack(property);
			} else if (property != message2[pname]) {
				returnVal = false;
				return;
			}
			propCount++;
		});
		
		forEach(message2, function(property, pname) {
			propCount--;
		});
		if(propCount != 0) {
			return false;
		}
		return returnVal;
	}
	
	
	//parse Functions
	
	function parsePrim (primNode, testWorld) {
		var primName = primNode.nodeName;
		if (primName == 'number') {
			var numAsString = primNode.getAttribute('value');
			return parseFloat(numAsString, 10);
		} else if (primName == 'sc') {
			var name = primNode.getAttribute('name');
			if (!testWorld.startCaps[name]) {
				return {
					scPatternMatch : scPatternMatch,
					callBack : function (inCap) {
						testWorld.startCaps[name] = inCap;
						lookupTable[name] = inCap;
					}
				};
			}
			return testWorld.startCaps[name];
		} else if (primName == 'undefined') {
			return undefined;
		}
		//TODO: add other primitive types
	}
	
	function parseStartCap (actionNode, testWorld) {
		var scname = actionNode.getAttribute('name');
		var sctype = actionNode.getAttribute('type');

		var sctypeFirstWord = sctype.split(" ")[0];

		//testWorld.startCaps[scname] = makeStartCap(sctypeFirstWord);
		if(sctypeFirstWord == 'assoc') {
			testWorld.startCaps[scname] = makeCellAssocInput();
		} else {
			testWorld.startCaps[scname] = makeCell();
		}
		
		testWorld.startCaps[scname].type = parseType(sctype);
		
		// make this more functional
		lookupTable[scname] = testWorld.startCaps[scname];
		return testWorld;
	}

	function parseEndCap (actionNode, testWorld) {
		var endCapName = actionNode.getAttribute('name');
		var expression = actionNode.getAttribute('expression');

		var scFromExp = evaluate(parseExpr(expression));

		testWorld.outMessages[endCapName] = [];

		scFromExp.injectFunc('output', null, function(keyVal) {
			if(keyVal.key != undefined && keyVal.val != undefined) {
				testWorld.outMessages[endCapName].push({action:'set', key:keyVal.key, value:keyVal.val});
			} else {
				testWorld.outMessages[endCapName].push({action:'set', key:keyVal});
			}
		});

		scFromExp.injectCustomRemoveLineResponse('output', function (keyVal, func, id) {
			if(keyVal.key != undefined && keyVal.val != undefined) {
				testWorld.outMessages[endCapName].push({action:'remove', key:keyVal.key});
			} else {
				testWorld.outMessages[endCapName].push({action:'remove', key:keyVal});
			}			
		});
		
		return testWorld;
	}
	
	function parseMessages (actionNode, testWorld) {
		forEach(children(actionNode), function (messageNode) {
			var messageName = messageNode.nodeName;
			var scName = messageNode.getAttribute('scname');
			
			var args = [];
			forEach(children(messageNode), function (childNode) {
				args.push(parsePrim(childNode, testWorld));
			});

			var value;
			if(args.length == 2) {
				value = {key:args[0], val:args[1]};
			} else if (args.length == 1) {
				value = args[0];
			}
			
			
			//testWorld.startCaps[scName].send([makeMessage[messageName].apply(null, args)]);
			if (messageName == "set") {
				testWorld.startCaps[scName].addLine(value);
			} else if (messageName == "remove") {
				testWorld.startCaps[scName].removeLine(value);
			}
		});
		return testWorld;
	}
	
	function messageToString (message, ecName) {
		if(message.value) {
			return message.action + "(" + message.key + ", " + message.value + ") at endCap: " + ecName;
		} else {
			//return message.action + "(" + message.value + ") at endCap: " + ecName;
			return message.action + "(" + message.key + ") at endCap: " + ecName;
		}
	}
	
	function parseExpectedMessages (actionNode, testWorld) {
		forEach(children(actionNode), function (messageNode) {
			var messageName = messageNode.nodeName;
			var ecName = messageNode.getAttribute('ecname');

			var childNodes = children(messageNode);
			var args = [];
			forEach(children(messageNode), function (childNode) {
				args.push(parsePrim(childNode, testWorld));
			});
			var messageToCheck;
			if(args.length == 2) {
				messageToCheck = {action:messageName, key:args[0], value:args[1]};
			} else if (args.length == 1) {
				messageToCheck = {action:messageName, key:args[0]};
			}
			
			if (testWorld.outMessages[ecName].length == 0) {
				testWorld.testOutput += "Error: Expected Message: " + messageToString(messageToCheck, ecName);
				testWorld.testOutput += ", but received NO Message\n";
			}
			else if (messagesEqual(testWorld.outMessages[ecName][0], messageToCheck)) {
				testWorld.testOutput += "Confirmed Message: " + messageToString(messageToCheck, ecName) + "\n";
				testWorld.outMessages[ecName].shift();
			} else {
				testWorld.testOutput += "Error: Expected Message: " + messageToString(messageToCheck, ecName);
				testWorld.testOutput += ", but received Message: " + messageToString(testWorld.outMessages[ecName][0], ecName) + "\n";
			}
		});
		return testWorld;
	}
	
	function parseRemoveCap (actionNode, testWorld) {
		return testWorld;
	}

	
	function parseAndRun(xml) {
		var testNodes = xpath("test", xml);

		forEach(testNodes, function (testNode) {
			var testWorld = {startCaps:{}, endCaps:{}, outMessages:[], testOutput:""};
			forEach(children(testNode), function (actionNode) {
				var action = actionNode.nodeName;
				if (action == 'startcap') {
					testWorld = parseStartCap(actionNode, testWorld);
				} else if (action == 'endcap') {
					testWorld = parseEndCap(actionNode, testWorld);
				} else if (action == 'messages') {
					testWorld = parseMessages(actionNode, testWorld);
				} else if (action == 'expectedmessages') {
					testWorld = parseExpectedMessages(actionNode, testWorld);
				} else if (action == 'removecap') {
					testWorld = parseRemoveCap(actionNode, testWorld);
				}
			});
			var testName = testNode.getAttribute("name");
			console.log("Test", testName, "complete. Results: ");
			console.log(testWorld.testOutput);
		});
	}

	
	
	return function () {
		var xml = loadXMLNow(filename);
		parseAndRun(xml);
	}();
}

File: server/mbrella/v1.4/src/testsyntax.erl
-module(testsyntax).
-compile(export_all).
-include("../include/scaffold.hrl").
-include_lib("xmerl/include/xmerl.hrl").

-define( trace(X), io:format("TRACE ~p:~p ~p~n", [?MODULE, ?LINE, X])).

-record(testWorld, {startCaps=dict:new(), endCaps=dict:new(), outMessages=dict:new()}).
% 
% parseAll(D) ->
% 	% find all RSS files underneath D
% 	FL = filelib:fold_files(D, ".+.rss$", true, fun(F, L) -> [F|L] end, []),
% 	[ parse(F) || F <- FL ].


parse(FName) ->
	% parses a single RSS file
	{R,_} = xmerl_scan:file(FName),
	spawn(fun() -> 
		L = extract(R,undefined)
	end).
	%L = lists:reverse(extract(R, [])),
	% print channel title and data for first two episodes
	%io:format("~n>> ~p~n", [element(1,lists:split(3,L))]),
	%L,

% handle 'xmlElement' tags
extract(R, State) when is_record(R, xmlElement) ->
	case R#xmlElement.name of
		tests ->
			lists:foldl(fun extract/2, State, R#xmlElement.content);
		test ->
			io:format("Running test ~p~n", [getAtt(name, R)]),
			lists:foldl(fun extract/2, #testWorld{}, R#xmlElement.content);
		startcap ->
			parseStartCap(R, State);
		endcap ->
			parseEndCap(R, State);
		messages ->
			parseMessages(R, State);
		expectedmessages ->
			parseExpectedMessages(R, State);
		removeCap ->
			parseRemoveCap(R, State);
		_ -> % for any other XML elements, simply iterate over children
			lists:foldl(fun extract/2, State, R#xmlElement.content)
	end;
extract(#xmlText{}, State) -> State.  % ignore any other text data


parseStartCap(R, State) ->
	Name = getAtt(name, R),
	Type = getAtt(type, R),
	#testWorld{startCaps=StartCaps} = State,
	SCTypeFirstWord = hd(string:tokens(Type, " ")),
	SC = if
		SCTypeFirstWord == "map" -> 
			Result = cell:makeCellMapInput(), 
			Cell = Result#exprCell{type=type:parse(Type)},
			env:nameAndStoreCell(Cell);
		true -> 
			Result = cell:makeCell(),
			Cell = Result#exprCell{type=type:parse(Type)},
			env:nameAndStoreCell(Cell)
	end,
	NewStartCaps = dict:store(Name, SC, StartCaps),
	State#testWorld{startCaps=NewStartCaps}.

parseEndCap(R, State) ->
	#testWorld{startCaps=StartCaps, outMessages=OutMessages} = State,
	Name = getAtt(name, R),
	Expression = getAtt(expression, R),
	EC = eval:evaluate(expr:customExprParse(Expression, StartCaps)),

	Self = self(),
	cell:injectFunc(EC, fun(Val) -> 
		Self ! {response, Name, {set, Val}},
		fun() -> 
			case Val of
				{K,V} -> Self ! {response, Name, {remove, K}};
				_ -> Self ! {response, Name, {remove, Val}}
			end
		end
	end),
	State#testWorld{outMessages=dict:store(Name, [], OutMessages)}.
	
%private helper function
parseArgs(R) ->
	GetArg = fun(InnerR, L) ->
		case parsePrim(InnerR) of
			undefined -> L;
			Prim -> [Prim | L]
		end
	end,
	Result = lists:foldr(GetArg, [], R#xmlElement.content),
	case Result of
		[A,B|[]] -> {A,B};
		[A|[]] -> A;
		[] -> noArgs
	end.

extractMessage(R, State) when is_record(R, xmlElement) ->
	#testWorld{startCaps=StartCaps} = State,
	GetStartCap = fun() ->
		SCName = getAtt(scname, R),
		dict:fetch(SCName, StartCaps)
	end,
	Args = parseArgs(R),
	Value = case Args of 
		{scPatternMatch, Name} -> 
			dict:fetch(Name, StartCaps);
		{Key, {scPatternMatch, Name}} ->
			{Key, dict:fetch(Name, StartCaps)};
		{func, Name} ->
			env:lookup(Name);
		{Key, {func, Name}} ->
			{Key, env:lookup(Name)};			
		Arg -> Arg
	end,
	case R#xmlElement.name of
		set ->
			StartCap = GetStartCap(),
			cell:addLine(StartCap, Value);
		remove ->
			StartCap = GetStartCap(),
			cell:removeLine(StartCap, Value);
		_ -> undefined
	end;
extractMessage(_, _) -> undefined.  % ignore any other text data

	
parseMessages(R, State) ->
	lists:map(fun(InnerR) -> extractMessage(InnerR, State) end, R#xmlElement.content),
	State.
	
	
	
parseExpectedMessages(R, State) ->
	ExpectedMessages = lists:foldl(fun(InnerR, InnerExpectedMessages) -> extractExpectedMessage(InnerR, InnerExpectedMessages) end, [], R#xmlElement.content),
	loop(ExpectedMessages, [], State).

extractExpectedMessage(R, ExpectedMessages) when is_record(R, xmlElement) ->
	ECName = getAtt(ecname, R),
	MessageToCheck = {R#xmlElement.name, parseArgs(R)},
	ExpectedMessages ++ [{ECName, MessageToCheck, {false, undefined}}];
extractExpectedMessage(_, State) -> State.  % ignore any other text data

loop(ExpectedMessages, FinishedMessages, State) ->
	receive
		{response, ECName, Message} ->
			NewExpectedMessages = lists:map(fun(EMessage) -> checkIncomingMessage(EMessage, Message, State) end, ExpectedMessages),
			NewerState = lists:foldl(fun updateState/2, State, NewExpectedMessages),
			NewerExpectedMessages = lists:filter(fun filterExpectedMessages/1, NewExpectedMessages),
			NewFinishedMessages = FinishedMessages ++ lists:filter(fun filterFinishedMessages/1, NewExpectedMessages),
			if
				length(NewerExpectedMessages) == 0 -> 
					lists:map(fun outputExpectedMessage/1, NewFinishedMessages),
					NewerState;
				true -> loop(NewerExpectedMessages, NewFinishedMessages, NewerState)
			end
	after 1000 ->
		io:format("Test Timed Out~n", []),
		lists:map(fun outputExpectedMessage/1, FinishedMessages),
		lists:map(fun outputExpectedMessage/1, ExpectedMessages),
		State
	end.
	
filterExpectedMessages({_,_,{false,_}}) -> true;
filterExpectedMessages(_) -> false.
	
filterFinishedMessages({_,_,{true,_}}) -> true;
filterFinishedMessages(_) -> false.

scMatch(MessageToCheck, TrySC, StartCaps) ->
	case MessageToCheck of
		{scPatternMatch, TrySCName} -> 
			try dict:fetch(TrySCName, StartCaps) of
				TrySC -> match;
				_ -> false
			catch
				_:_ -> {scPatternMatch, TrySCName, TrySC}
			end;
		_ -> false
	end.	
	
checkIncomingMessage({ECName, MessageToCheck, {true, Output}}, IncomingMessage, State) -> {ECName, MessageToCheck, {true, Output}};
checkIncomingMessage({ECName, MessageToCheck, {false, _}}, IncomingMessage, State) ->
	#testWorld{startCaps=StartCaps} = State,
	Response =
		case IncomingMessage of
			MessageToCheck -> match;
			{Action, {Key,TrySC}} ->
				case MessageToCheck of
					{Action, {Key, MVal}} -> 
						case scMatch(MVal, TrySC, StartCaps) of
							false -> {badMatch, IncomingMessage};
							SCMatch -> SCMatch
						end;
					_ -> {badMatch, IncomingMessage}
				end;
			{Action, TrySC} ->
				case MessageToCheck of
					{Action, MVal} -> 
						case scMatch(MVal, TrySC, StartCaps) of
							false -> {badMatch, IncomingMessage};
							SCMatch -> SCMatch
						end;
					_ -> {badMatch, IncomingMessage}
				end;
			BadMatch -> {badMatch, BadMatch}
		end,
	Result = case Response of
		match -> {true, {match, "Confirmed Message: " ++ messageToString(MessageToCheck, ECName)}};
		none -> {true, {none, "Confirmed Message: " ++ messageToString(MessageToCheck, ECName)}};
		{scPatternMatch, SCName, SC} -> {true, {{scPatternMatch, SCName, SC}, "Confirmed Message: " ++ messageToString(MessageToCheck, ECName)}};
		{badMatch, WrongMessage} -> {false, "Error: Expected Message: " ++ messageToString(MessageToCheck, ECName) ++ ", but received message " ++ messageToString(WrongMessage, ECName)};
		noMessage -> {false, "Error: Expected Message: " ++ messageToString(MessageToCheck, ECName) ++ ", but received NO Message"}
	end,
	{ECName, MessageToCheck, Result}.


updateState({ECName, MessageToCheck, {true, {Control, _}}}, State) ->
	#testWorld{startCaps=StartCaps} = State,
	case Control of
		match -> State;
		none -> State;
		{scPatternMatch, SCName, SC} -> State#testWorld{startCaps=dict:store(SCName, SC, StartCaps)}	
	end;
updateState(_, State) ->
	State.
	
outputExpectedMessage({ECName, MessageToCheck, {true, Output}}) ->
	{Control, Message} = Output,
	io:format("~p~n", [Message]);
outputExpectedMessage({ECName, {none, _}, _}) ->
	io:format("No message Expected at endCap: ~p~n", [ECName]);
outputExpectedMessage({ECName, MessageToCheck, {false, Output}}) ->
	io:format("~p~n", [Output]).
	

toString(Value) ->
	case Value of
		List when is_list(List) -> List;
		Num when is_integer(Num) -> integer_to_list(Num);
		Num when is_float(Num) -> float_to_list(Num);
		Atom when is_atom(Atom) -> atom_to_list(Atom);
		{scPatternMatch, SCName} -> "StartCap-" ++ SCName;
		{func, FuncName} -> "Function-" ++ FuncName;
		_ -> "No Match for this String"
	end.

messageToString({Name, {Key, Value}}, ECName) ->
	toString(Name) ++ "(" ++ toString(Key) ++ ", " ++ toString(Value) ++ ") at endCap: " ++ ECName;
messageToString({Name, Value}, ECName) ->
	toString(Name) ++ "(" ++ toString(Value) ++ ") at endCap: " ++ ECName.

	
parseRemoveCap(R, State) ->
	State.


parsePrim(R) when is_record(R, xmlElement) ->
	case R#xmlElement.name of
		number -> list_to_integer(getAtt(value, R));
		bool -> list_to_atom(getAtt(value, R));
		func -> {func, getAtt(name, R)};
		sc -> {scPatternMatch, getAtt(name, R)};
		_ -> undefined
	end;
parsePrim(_) ->
	undefined.

fFunc(Name) -> fun(X) -> X#xmlAttribute.name == Name end.
getAtt(Name, R) -> (hd(lists:filter(fFunc(Name), R#xmlElement.attributes)))#xmlAttribute.value.


% 'main' function (invoked from shell, receives command line arguments)
main() ->
	%D = atom_to_list(hd(A)),
	%parseAll(D).
	parse("../../../shared/testing/erlangTests.xml").